pkg/ignition/ignition.go

```diff
@@ -0,0 +1,134 @@
+package ignition
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	ignCfg "github.com/coreos/ignition/v2/config/v3_2"
+	ignCfgTypes "github.com/coreos/ignition/v2/config/v3_2/types"
+	mcfg "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
+	"github.com/pkg/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+//+kubebuilder:rbac:groups="machineconfiguration.openshift.io",resources=machineconfigs,verbs=list
+
+const (
+	// kubeletSystemdName is the name of the systemd service that the kubelet runs under,
+	// this is used to parse the kubelet args
+	kubeletSystemdName = "kubelet.service"
+	// CloudConfigOption is the kubelet CLI option for the cloud configuration file
+	CloudConfigOption = "cloud-config"
+	// CloudProviderOption is the kubelet CLI option for cloud provider
+	CloudProviderOption = "cloud-provider"
+	// RenderedWorkerPrefix allows identification of the rendered worker MachineConfig, the combination of all worker
+	// MachineConfigs.
+	RenderedWorkerPrefix = "rendered-worker-"
+)
+
+// Ignition is a representation of an Ignition file
+type Ignition struct {
+	config ignCfgTypes.Config
+}
+
+// New returns a new instance of Ignition
+func New(c client.Client) (*Ignition, error) {
+	log := ctrl.Log.WithName("ignition")
+	machineConfigs := &mcfg.MachineConfigList{}
+	err := c.List(context.TODO(), machineConfigs)
+	if err != nil {
+		return nil, err
+	}
+	renderedWorker, err := getLatestRenderedWorker(machineConfigs.Items)
+	if err != nil {
+		return nil, err
+	}
+
+	configuration, report, err := ignCfg.Parse(renderedWorker.Spec.Config.Raw)
+	if err != nil || report.IsFatal() {
+		return nil, errors.Errorf("failed to parse MachineConfig ignition: %v\nReport: %v", err, report)
+	}
+	ign := &Ignition{
+		config: configuration,
+	}
+	log.V(1).Info("parsed", "machineconfig", renderedWorker.GetName(), "ignition version",
+		configuration.Ignition.Version)
+	return ign, nil
+}
+
+// GetKubeletArgs returns a set of arguments for kubelet.exe, as specified in the ignition file
+func (ign *Ignition) GetKubeletArgs() (map[string]string, error) {
+	var kubeletUnit ignCfgTypes.Unit
+	for _, unit := range ign.config.Systemd.Units {
+		if unit.Name == kubeletSystemdName {
+			kubeletUnit = unit
+			break
+		}
+	}
+	if kubeletUnit.Contents == nil {
+		return nil, errors.Errorf("ignition missing kubelet systemd unit file")
+	}
+	argsFromIgnition, err := parseKubeletArgs(*kubeletUnit.Contents)
+	if err != nil {
+		return nil, errors.Wrap(err, "error parsing kubelet systemd unit args")
+	}
+	return argsFromIgnition, nil
+}
+
+// getLatestRenderedWorker returns the most recently created rendered worker MachineConfig
```

Comment:
We create some of the bootstrap files from the contents of the rendered-worker MachineConfig (ignition). We want to always grab the latest one so that if something changes in the ignition spec, we can pick these changes up when configuring next

Reply:
I suggest capturing the "why` in the comment also.